pkg/publish.go

```diff
@@ -250,3 +260,83 @@ func getBytesSha256(b []byte) string {
 	h.Write(b)
 	return hex.EncodeToString(h.Sum(nil))
 }
+
+func getAnchorHrefsFromHTML(content string) ([]string, error) {
+	var anchorHrefs []string
+	z := html.NewTokenizer(strings.NewReader(content))
+	for {
+		tokenType := z.Next()
+		token := z.Token()
+
+		switch tokenType {
+		case html.ErrorToken:
+			if z.Err() != io.EOF {
+				return nil, z.Err()
+			} else {
+				return anchorHrefs, nil
+			}
+		case html.StartTagToken:
+			if token.Data == "a" {
+				for _, attr := range token.Attr {
+					if attr.Key == "href" {
+						anchorHrefs = append(anchorHrefs, attr.Val)
+					}
+				}
+			}
+		}
+	}
+	return anchorHrefs, nil
+}
+
+func getLocalLinkedAnchors(content string, file string) ([]string, error) {
+	fileDir := filepath.Dir(file)
+	anchorHrefs, err := getAnchorHrefsFromHTML(content)
+	if err != nil {
+		return nil, err
+	}
+
+	var localAnchorHrefs []string
+
+	for _, anchorHref := range anchorHrefs {
+		// skip anchorHrefs that are URLs
+		if _, err := url.ParseRequestURI(anchorHref); err == nil {
+			continue
+		}
+
+		anchorHrefPath := filepath.Join(fileDir, anchorHref)
+		if _, err := os.Stat(anchorHrefPath); !os.IsNotExist(err) {
+			localAnchorHrefs = append(localAnchorHrefs, anchorHref)
+		} else {
+			fmt.Printf("warn: could not find file %s\n", anchorHrefPath)
+		}
+	}
+
+	return localAnchorHrefs, nil
+}
+
+func replaceRelativeLinks(file string, pageContent string, uri string, repoBrowseURL string, repoRoot string) (string, error) {
+	localAnchorHrefs, err := getLocalLinkedAnchors(pageContent, file)
+	if err != nil {
+		return "", err
+	}
+
+	fileDir := filepath.Dir(file)
+	for _, localAnchorHref := range localAnchorHrefs {
+		localAnchorHrefPath := filepath.Join(fileDir, localAnchorHref)
+
+		src, err := source.New(localAnchorHrefPath, source.Opts{StripComments: true, TrimSpace: true})
+		if err != nil {
+			// file exists but is not a dox source file, link to stash instead
+			localAnchorHrefFromRepoRoot := strings.Replace(localAnchorHrefPath, repoRoot, "", -1)
+			if strings.HasPrefix(localAnchorHrefFromRepoRoot, "/") {
+				localAnchorHrefFromRepoRoot = localAnchorHrefFromRepoRoot[1:]
+			}
+			pageContent = strings.Replace(pageContent, fmt.Sprintf(`href="%s"`, localAnchorHref), fmt.Sprintf(`href="%s/%s"`, repoBrowseURL, localAnchorHrefFromRepoRoot), -1)
+			continue
+		} else if src.ID() != "" {
+			pageContent = strings.Replace(pageContent, fmt.Sprintf(`href="%s"`, localAnchorHref), fmt.Sprintf(`href="%s/pages/viewpage.action?pageId=%s"`, uri, src.ID()), -1)
```

Comment:
It doesn't feel right to have Confluence specific URL patterns embedded as a literal. I think it should be a const at least. I wouldn't mind making `dox` able to publish to Github Wiki or other targets in time, so the more we can keep from sprinkling target specific details all over, the easier it will be to abstract them into interfaces.

Reply:
- [ ] address Confluence URL